rfc 0015, part 1 - warn on use of ' as :: in identifiers

tonycoz opened the pull request:
This implements the first part of [RFC 0015](https://github.com/Perl/RFCs/blob/main/rfcs/rfc0015.md), which states we should warn on the use of ' instead of :: in identifiers.

tonycoz committed:
Enable the current "Old package separator used in string" by default

If we're going to be removing this entirely, the behavior of
existing code is going to change.  Make sure everyone sees it.

tonycoz committed:
toke.c: deprecation warning for ' as a package separator

First stage of RFC 0015.

This also changes the warning for ' as package separator in
quoted strings to also be a deprecation warning.

rjbs commented:
Hey, nice.  I am not qualified to code review this, but am qualified to say "thanks!"

nicomen commented:
Nice, I never saw a reply on how to submit my work in the RFC itself Perl/RFCs#18, but my changes related to this are here if you want to take a look https://github.com/Perl/perl5/compare/blead...nicomen:perl5:nicomen/warn_on_old_package_separator. I've been struggling with how to get a proper warning (suggesting the correct syntax) for sub names as the token contains a space.

tonycoz commented:
> Nice, I never saw a reply on how to submit my work in the RFC itself

Sorry, I didn't see that.  I think the best way to get a code review is to make a PR, if you don't think it's ready to merge, make it draft.

abraxxa commented on pod/perldiag.pod:
```diff
@@ -4502,9 +4502,15 @@ C<sysread()>ing a file, or when seeking past the end of a scalar opened
+
+(W deprecated, syntax) You used the old package separator "'" in a
+variable, subroutine or package name.  Support for the old package
+separator will be removed in Perl 5.38.
```
Shouldn't this say '5.40' as it's deprecated in 5.38 but removed according to the deprecation policy?

tonycoz replied:
Thanks, fixed.

tonycoz commented:
Some updates:
- changed the removal version to 5.40
- added a removal notice to the ' separator in quoted string warning
- suppressed the warnings in two more core tests

A few bundled cpan modules also warn:
```
autodie:

Old package separator "'" deprecated at t/exception_class.t line 46.
Old package separator "'" deprecated at /home/tony/dev/perl/git/perl6/cpan/autodie/t/lib/pujHa/ghach.pm line 1.
Old package separator "'" deprecated at /home/tony/dev/perl/git/perl6/cpan/autodie/t/lib/pujHa/ghach/Dotlh.pm line 1.

../cpan/Scalar-List-Utils/t/blessed.t ................................ ok
../cpan/Scalar-List-Utils/t/dualvar.t ................................ ok
Old package separator "'" deprecated at (eval 285) line 1.
../cpan/Scalar-List-Utils/t/exotic_names.t ........................... ok
../cpan/Scalar-List-Utils/t/first.t .................................. ok
../cpan/Scalar-List-Utils/t/getmagic-once.t .......................... ok


io/socket.t .......................................................... ok
Old package separator "'" deprecated at test.pl line 124.
../cpan/Term-Cap/test.pl ............................................. ok

Old package separator "'" deprecated at t/parent-classfromclassfile.t line 21.
../cpan/parent/t/parent-classfromclassfile.t ......................... ok
```

I'll make PRs (or tickets) against upstreams once (if) this is merged.

nicomen commented:
Please look at the RFC for some already reported upstream bugs/pull requests that may need a nudge.

leonerd commented on t/comp/package.t:
```diff
@@ -28,10 +36,13 @@ $ABC = join(':', sort(keys %ABC::));
     print $xyz eq 'bar:main:new:xyz:ABC' ? "ok 1\n" : "not ok 1 '$xyz'\n";
 } else {
-    print $xyz eq 'ABC:bar:main:new:xyz' ? "ok 1\n" : "not ok 1 '$xyz'\n";
+    print $xyz eq 'ABC:BEGIN:bar:main:new:xyz' ? "ok 1\n" : "not ok 1 '$xyz'\n";
```
FYI: This behaviour of `BEGIN` blocks really annoys me. I have vague plans to fix it one day...

leonerd commented on t/op/ref.t:
```diff
@@ -265,8 +265,10 @@ is (join('', sort values %$anonhash2), 'BARXYZ');
-
-$object = bless $main'anonhash2;
+{
+    no warnings "syntax", "deprecated";
```
Should be `qw(...)` to match the other tests.

tonycoz replied:
Fixed.

leonerd commented on t/op/ref.t:
```diff
@@ -290,7 +292,10 @@ sub mymethod {
 $string = "good";
-$main'anonhash2 = "foo";
+{
+    no warnings "syntax", "deprecated";
```
Ditto `qw(...)` here and further below

tonycoz replied:
Fixed in a few places here, and in some other tests.

leonerd reviewed (commented):
Overall seems good, give or take some minor formatting comments in the tests.

As this is labelled "Part 1" I imagine it's not complete for the RFC yet. What further parts are required?

tonycoz commented:
> Overall seems good, give or take some minor formatting comments in the tests.
> 
> As this is labelled "Part 1" I imagine it's not complete for the RFC yet. What further parts are required?

The RFC calls them "steps":

https://github.com/Perl/RFCs/blob/main/rfcs/rfc0015.md#specification

and they belong in separate releases.

The first step is this, effectively deprecating ' in identifiers.

The second step is turning them into a syntax error, though it's unclear whether that should be a "directly produce a syntax error where `'` is now accepted" or "don't accept `'` in identifiers, the `'` is part of the next token, which may produce a syntax error".

demerphq commented:
Is there anything needed on this still? It looks good to me.

leonerd reviewed (approved):
LGTM.

Can we get this in in time for 5.37.9?

demerphq commented:
@tonycoz for the win. Thanks!
